### PR TITLE
DrivAerML: relative L2 training loss (align with eval metric)

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

DrivAerML trains on absolute MSE but evaluates on `surface_rel_l2_pct` — a relative (normalized) L2. This objective mismatch means gradient signal is dominated by high-magnitude cases, but the evaluation weights all cases equally. Using relative L2 during training should align gradient signal with the evaluation criterion.

**Formula:** For each node: `rel_l2 = ||pred - target||_2 / (||target||_2 + eps)`. Mean over nodes and cases.

## Instructions

Implement `--loss rel_l2` option in train.py. When active, replace the MSE loss with:

```python
def relative_l2_loss(pred, target, eps=1e-8):
    # pred, target: [N, C] or [N] per case
    num = (pred - target).norm(dim=-1)  # per-node L2 error
    denom = target.norm(dim=-1) + eps    # per-node target norm
    return (num / denom).mean()
```

For DrivAerML, `target` is `surface_cp` (1 channel). Apply normalization per node.

Also try: `--loss mixed` = 0.5 * MSE + 0.5 * rel_l2.

Run two variants:
```bash
# Variant 1: Pure relative L2
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 \
  --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 \
  --model-heads 8 --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --loss rel_l2 --wandb_group dm-loss-alignment

# Variant 2: Mixed 50/50
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 \
  --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 \
  --model-heads 8 --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --loss mixed --wandb_group dm-loss-alignment
```

**Divergence check:** If rel_l2 diverges before epoch 20, add `--grad-clip 1.0` and document.

## Reporting Requirements
- Best val `surface_rel_l2_pct` for each variant
- Test at best val checkpoint
- W&B run IDs for both variants

## Baseline

**DrivAerML baselines:**
- val best: `3.997%` @ ep467 (PR #2898, W&B: `bht6h42t`, 4L/512d, AdamW lr=5e-4, T_max=30)
- test best: `6.244%` (PR #2648, 4L/320d, truncated eval — OLDER CONFIG)
- External targets: AB-UPT 3.82% | Transolver-3: 3.71%

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT fern:

## Results

### Summary

Both relative L2 loss variants **failed catastrophically** — the model gets trapped at a constant-prediction plateau (~75% rel_l2_pct) and never learns, while the baseline MSE loss reaches 3.997%.

### Variant 1: Pure Relative L2 Loss (`--loss rel_l2`)

| Metric | Value |
|--------|-------|
| Best val `surface_rel_l2_pct` | **75.27%** @ ep138 |
| Test `surface_rel_l2_pct` (best val ckpt) | **76.21%** |
| Final test `surface_rel_l2_pct` | 76.38% |
| Epochs completed | 505 |
| Peak memory | ~6.2 GB |
| W&B run ID | `lago3za1` |

**Command:**
```bash
CUDA_VISIBLE_DEVICES=0 SENPAI_MAX_EPOCHS=999 SENPAI_TIMEOUT_MINUTES=360.0 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 \
  --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 \
  --model-heads 8 --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --loss rel_l2 --wandb-group dm-loss-alignment \
  --agent fern --wandb-name "fern/dm-rel-l2-loss"
```

### Variant 2: Mixed 50/50 Loss (`--loss mixed`)

| Metric | Value |
|--------|-------|
| Best val `surface_rel_l2_pct` | **75.30%** @ ep328 |
| Test `surface_rel_l2_pct` (best val ckpt) | **76.30%** |
| Final test `surface_rel_l2_pct` | 76.37% |
| Epochs completed | 498 |
| Peak memory | ~6.0 GB |
| W&B run ID | `cxrfryjm` |

**Command:**
```bash
CUDA_VISIBLE_DEVICES=1 SENPAI_MAX_EPOCHS=999 SENPAI_TIMEOUT_MINUTES=360.0 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 \
  --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 \
  --model-heads 8 --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --loss mixed --wandb-group dm-loss-alignment \
  --agent fern --wandb-name "fern/dm-mixed-loss"
```

### Baseline Comparison

| Variant | Best val rel_l2_pct | Best test rel_l2_pct |
|---------|--------------------|--------------------|
| **Baseline MSE** (PR #2898) | **3.997%** @ ep467 | **6.244%** |
| Pure rel_l2 | 75.27% @ ep138 | 76.21% |
| Mixed 50/50 | 75.30% @ ep328 | 76.30% |

### What happened

**Both variants are ~19x worse than the baseline.** The model converges to a near-constant prediction (the z-normalized mean) within the first 5 epochs and stays trapped there for the remaining 500 epochs. Training loss does decrease, but the validation metric never improves beyond ~75%.

**Root cause — relative L2 on z-normalized targets creates a degenerate loss landscape:**

1. The training loss operates on z-normalized targets (mean≈0, std≈1). Many surface nodes have z-target ≈ 0.
2. The relative L2 formula `|pred - target| / (|target| + eps)` with eps=1e-8 creates enormous gradients for near-zero targets — effectively infinite weight on these nodes.
3. The model quickly learns to predict near-zero everywhere (the z-space mean), which minimizes the huge penalties from near-zero denominator nodes.
4. Once trapped at this "predict the mean" solution, there's no gradient signal to escape — the relative error is uniformly ~75% across all cases.
5. Even the mixed loss (50% MSE + 50% rel_l2) suffers: the rel_l2 component dominates the gradient landscape and prevents the MSE term from driving learning.

**Divergence check:** No NaN/Inf divergence occurred. The rel_l2 training loss was very noisy (oscillating 1.0-17.0) with high grad norms (300-700), but technically stable. Grad clipping was not added since there was no divergence per the PR criterion, though the model simply failed to learn rather than diverging.

### Suggested follow-ups

1. **Apply relative L2 in raw (un-normalized) space** — compute the loss after `transform.invert()`, matching how the eval metric is computed. This avoids the near-zero denominator problem since raw Cp values are typically O(1).
2. **Use a much larger eps** (e.g., eps=1.0 or eps=mean(|target|)) to prevent near-zero targets from dominating.
3. **Curriculum approach** — start with MSE and gradually blend in rel_l2 after the model has learned reasonable predictions, avoiding the early constant-prediction trap.
4. **Log-cosh relative loss** — use `log(cosh(rel_error))` instead of raw `rel_error` to bound the gradient contribution from near-zero denominator nodes.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #3074 because ADVISOR: Both pure rel_l2 (75.27%) and 50/50 mixed (75.30%) immediately converge to the z-normalized mean and plateau there for the entire run — a degenerate failure. Root cause: relative L2 on z-normalized targets creates infinite-weight gradients at near-zero nodes, driving the model to predict zero. The loss formulation and feature normalization are fundamentally incompatible as-is. The fix (compute rel_l2 in raw un-normalized space) would require a different implementation and is a distinct hypothesis. Closing..
